src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/params.py

```diff
@@ -3,3 +3,25 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
+from ._util import ParametersContext, patch_arg_make_required, patch_arg_update_description
+
+
+with ParametersContext(command='sql server create') as c:
+    from azure.mgmt.sql.models.server import Server
+
+    # About the patches:
+    #
+    # - Both administrator_login and administrator_login_password are required for server creation.
+    # However these two parameters are given default value in the create_or_update function
+    # signature, therefore, they can't be automatically converted to requirement arguments.
+    #
+    # - Due to issue https://github.com/Azure/azure-cli/issues/1644, the description of version
+    # can't be correctly extracted
+    #
+    c.expand('parameters', Server, patches={
+        'administrator_login': patch_arg_make_required,
```

Comment:
I try to find a better way to automatically register parameters. The mechanism here is to automatically expand the 'parameter' argument in the sdk function signature base on the type of the object is supposed to be fed and add parameters to our command automatically.

Ideally this line should looks like:
```
    c.expand('parameters', Server)
```
It means, this command is wired to one SDK method, and the method's `parameter` accept a complex type of `Server` as its value. Automatically inspect the type `Server` and convert it's required initialize properties to the parameters on to the CLI command.

I find this approach to be more readable and concise. The patches here are used to address the information which CLI is not able to retrive from the SDK method as I explained in the comments. 

Yes this is a deviating from the command definition norm in other command module, but I hope I can demote that this is a better approach. For example, here is the full `commands.py` after I added all the commands from SQL SDL:
```
from ._util import (get_sql_servers_operation, get_sql_database_operations,
                    get_sql_elasticpools_operations, get_sql_recommended_elastic_pools_operations,
                    ServiceGroup, create_service_adapter)

server_operations = create_service_adapter('azure.mgmt.sql.operations.servers_operations',
                                           'ServersOperations')

with ServiceGroup(__name__, get_sql_servers_operation, server_operations) as s:
    with s.group('sql server') as c:
        c.command('create', 'create_or_update')
        c.command('delete', 'delete')
        c.command('show', 'get_by_resource_group')
        c.command('usage', 'list_usages')
        c.command('list', 'list_by_resource_group')
        c.generic_update_command('update', 'get_by_resource_group', 'create_or_update')

    with s.group('sql server service-objective') as c:
        c.command('list', 'list_service_objectives')
        c.command('show', 'get_service_objective')

    with s.group('sql server firewall') as c:
        c.command('create', 'create_or_update_firewall_rule')
        c.command('update', 'create_or_update_firewall_rule')
        c.command('delete', 'delete_firewall_rule')
        c.command('show', 'get_firewall_rule')
        c.command('list', 'list_firewall_rules')

database_operations = create_service_adapter('azure.mgmt.sql.operations.databases_operations',
                                             'DatabasesOperations')

with ServiceGroup(__name__, get_sql_database_operations, database_operations) as s:
    with s.group('sql database') as c:
        c.command('create', 'create_or_update')
        c.command('show', 'get')
        c.command('list', 'list_by_server')
        c.command('usage', 'list_usages')
        c.command('delete', 'delete')
        c.generic_update_command('update', 'get', 'create_or_update')

    with s.group('sql database rl') as c:
        c.command('list', 'list_replication_links')
        c.command('show', 'get_replication_link')
        c.command('delete', 'delete_replication_link')
        c.command('failover', 'failover_replication_link')
        c.command('force-failover', 'failover_replication_link_allow_data_loss')

    with s.group('sql database dw') as c:
        c.command('pause', 'pause_data_warehouse')
        c.command('resume', 'resume_data_warehouse')

    with s.group('sql database rp') as c:
        c.command('list', 'list_restore_points')

    with s.group('sql database tpe') as c:
        c.command('create', 'create_or_update_transparent_data_encryption_configuration')
        c.command('get-configuration', 'get_transparent_data_encryption_configuration')
        c.command('list-activity', 'list_transparent_data_encryption_activity')

    with s.group('sql database sta') as c:
        c.command('list', 'list_service_tier_advisors')
        c.command('show', 'get_service_tier_advisor')

elasticpools_ops = create_service_adapter('azure.mgmt.sql.operations.elastic_pools_operations',
                                          'ElasticPoolsOperations')

with ServiceGroup(__name__, get_sql_elasticpools_operations, elasticpools_ops) as s:
    with s.group('sql elastic-pools') as c:
        c.command('create', 'create_or_update')
        c.command('delete', 'delete')
        c.command('show', 'get')
        c.command('list', 'list_by_server')
        c.generic_update_command('update', 'get', 'create_or_update')

    with s.group('sql elastic-pools database') as c:
        c.command('list', 'list_databases')
        c.command('list-activity', 'list_database_activity')
        c.command('show', 'get_database')


recommanded_elastic_pools_ops = \
    create_service_adapter('azure.mgmt.sql.operations.recommended_elastic_pools_operations',
                           'RecommendedElasticPoolsOperations')

with ServiceGroup(__name__, get_sql_recommended_elastic_pools_operations,
                  recommanded_elastic_pools_ops) as s:
    with s.group('sql elastic-pools recommended') as c:
        c.command('show', 'get')
        c.command('list', 'list')
        c.command('metrics', 'list_metrics')

    with s.group('sql elastic-pools recommended database') as c:
        c.command('show', 'get_databases')
        c.command('list', 'list_databases')
```

Using the context objects, the definitions are organized into groups mapping to their corresponding 1) SDK class; 2) command groups. The cascades of indentations forced by the `with` statement improves the readability and provider easier approach to seach for command definition. 

Same with parameters definition:
```
with ParametersContext(command='sql server create') as c:
    from azure.mgmt.sql.models.server import Server

    # About the patches:
    #
    # - Both administrator_login and administrator_login_password are required for server creation.
    # However these two parameters are given default value in the create_or_update function
    # signature, therefore, they can't be automatically converted to requirement arguments.
    c.expand('parameters', Server, group_name='Creating', patches={
        'administrator_login': patch_arg_make_required,
        'administrator_login_password': patch_arg_make_required
    })

with ParametersContext(command='sql database create') as c:
    from azure.mgmt.sql.models.database import Database

    c.expand('parameters', Database, group_name='Creating')
```

The paradigms are the same. Focus on readability and code organization. Under the hood, I'm still using the same `cli_command` and `register_cli_argument` they're not different essentially.